### PR TITLE
[FIX] web: ensure required readonly fields are not underlined in list view

### PR DESCRIPTION
**Pull request by kcv-odoo**
Before this PR when list view is in multi edit mode and there is a field
with required and readonly modifiers, the field is underlined, which is not
intended as the field is in readonly mode.

task-3000817

### CODE REVIEW

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello Kartik,
This change will unfortunately render the border on the input rather than on the cell:

**This PR**
image

**Expected behaviour**
image

thanks

**Comment by kcv-odoo** [comment]
> Hello Kartik, This change will unfortunately render the border on the input rather than on the cell:
> 
> **This PR** image
> 
> **Expected behaviour** image
> 
> thanks

Should be fine now!! Thanks :)

File: addons/web/static/src/views/list/list_renderer.scss [hunk 1]
```diff
@@ -372,12 +372,16 @@
...
+        &:not(:has(.o_readonly_modifier)) {
+            border-bottom: 1px solid $o-input-border-required;
+        }
     }
 
-    .o_data_row.o_selected_row > .o_data_cell.o_invalid_cell:not(.o_readonly_modifier) {
-        border-bottom: 1px solid $danger;
+    .o_data_row.o_selected_row > .o_data_cell.o_invalid_cell {
+        &:not(:has(.o_readonly_modifier)) {
+            border-bottom: 1px solid $o-input-border-required;
```
Comment (stefanorigano) [inline addons/web/static/src/views/list/list_renderer.scss:383]:
```suggestion
            border-bottom: 1px solid $danger;
```
;)

File: addons/web/static/src/views/list/list_renderer.scss [hunk 2]
```diff
@@ -372,12 +372,16 @@
         }
     }
 
-    .o_data_row.o_selected_row > .o_data_cell.o_required_modifier:not(.o_readonly_modifier) {
-        border-bottom: 1px solid $o-input-border-required;
+    .o_data_row.o_selected_row > .o_data_cell.o_required_modifier {
+        &:not(:has(.o_readonly_modifier)) {
```
Comment (stefanorigano) [inline addons/web/static/src/views/list/list_renderer.scss:376]: No need to split in 2 lines. You can keep it like it was before your very last push.

https://github.com/odoo/odoo/compare/6eabe42a72e1f45a91c3e22f45493de170e2c8e5..a0a858ba5a4ed6b8b1c7c457da4fc09b901ece2f

**Review by stefanorigano — CHANGES_REQUESTED** [review]

**Review by stefanorigano — APPROVED** [review]
LGTM, thanks @kcv-odoo 👍

**Review by brboi — CHANGES_REQUESTED** [review]

**Comment by pparidans** [comment]
:warning: Unsupported feature on Firefox (and pre15.4 iOS) :warning:

**Review by IT-Ideas — COMMENTED** [review]
Hello @kcv-odoo 🙂 
Thanks for the fix.
Could you rephrase the commit like this?
```text
[FIX] web: ensure required readonly fields are not underlined in list view

Before this commit when list view is in multi edit mode and there is a field
with required and readonly modifiers, the field is underlined, which is not
intended as the field is in readonly mode.

task-3000817

```

**Comment by IT-Ideas** [comment]
Hello @pparidans, @brboi 👋🏼 
Thanks for having pointed that out!
What do you suggest?

Thanks 🙂

File: addons/web/static/src/views/list/list_renderer.scss [hunk 3]
```diff
@@ -372,11 +372,11 @@
         }
     }
 
-    .o_data_row.o_selected_row > .o_data_cell.o_required_modifier:not(.o_readonly_modifier) {
+    .o_data_row.o_selected_row > .o_data_cell.o_required_modifier:not(:is(.o_readonly_modifier)) {
```
Comment (pparidans) [inline addons/web/static/src/views/list/list_renderer.scss:375]: `:is` != `:has`

In this case, you basically rewritten the same selector as before :shrug:

**Review by pparidans — CHANGES_REQUESTED** [review]

File: addons/web/static/src/views/list/list_renderer.scss (hunk 3 above)
Comment (kcv-odoo) [reply addons/web/static/src/views/list/list_renderer.scss]: I know, But `:is` is supported in Firefox and almost every browser that's why I change it And I also added a new commit to add `o_readonly_modifier` on parent(`td`) also same as `o_required_modifier` and `o_invalid_cell` so it will fix issue without changing scss.
could you check?
Thanks!! :slightly_smiling_face: 
cc: @IT-Ideas @brboi

**Review by pparidans — CHANGES_REQUESTED** [review]
Looks ok to me

Could you squash your commits and add a test for this, please?

**Comment by kcv-odoo** [comment]
> Looks ok to me
> 
> Could you squash your commits and add a test for this, please?

Done!! Thanks :)

**Comment by IT-Ideas** [comment]
@pparidans I guess you will `r+`.
Thanks to all for your help 🙂

File: addons/web/static/tests/views/list_view_tests.js [hunk 4]
```diff
@@ -15682,4 +15682,18 @@ QUnit.module("Views", (hooks) => {
...
+            type: "list",
+            resModel: "foo",
+            serverData,
+            arch: `
+                <list>
+                    <field name="foo" readonly="1"/>
+                </list>
+            `,
+        });
+        assert.hasClass(target.querySelectorAll(".o_data_cell")[0], "o_readonly_modifier");
```
Comment (pparidans) [inline addons/web/static/tests/views/list_view_tests.js:15697]: Well, actually your assert could have simply been added in the "editable list view: readonly fields cannot be edited" test case.

**Review by pparidans — APPROVED** [review]
LGTM :+1: 
@robodoo r+
